cunumeric/module.py

```diff
@@ -3242,6 +3242,88 @@ def diagonal(
     )
 
 
+@add_boilerplate("a", "val")
+def fill_diagonal(
+    a: ndarray, val: ndarray, wrap: Optional[bool] = False
+) -> None:
+    """
+    Fill the main diagonal of the given array of any dimensionality.
+
+    For an array a with a.ndim >= 2, the diagonal is the list of locations with
+    indices a[i, ..., i] all identical. This function modifies the input
+    array in-place, it does not return a value.
+
+    Parameters
+    ----------
+
+    a : array, at least 2-D.
+        Array whose diagonal is to be filled, it gets modified in-place.
+    val : scalar or array_like
+        Value(s) to write on the diagonal. If val is scalar, the value is
+        written along the diagonal.
+        If array-like, the flattened val is written along
+        the diagonal, repeating if necessary to fill all diagonal entries.
+    wrap : bool
+        For tall matrices (2d) the diagonal “wrapped” after N columns.
+        You can have this behavior with this option. This affects
+        only tall matrices.
+
+    Raises
+    ------
+    ValueError
+        If the dimension of `a` is less than 2.
+
+    Notes
+    -----
+
+    See Also
+    --------
+    numpy.fill_diagonal
+
+    Availability
+    --------
+    Multiple GPUs, Multiple CPUs
+
+    """
+    if a.ndim < 2:
+        raise ValueError("array must be at least 2-d")
+
+    n = _builtin_min(a.shape)
+
+    if a.ndim > 2:
+        for s in a.shape:
+            if s != n:
+                raise ValueError(
+                    "All dimensions of input must be of equal length"
+                )
+
+        if wrap:
+            raise ValueError("wrap can be set to True only for tall matrices")
+
+    idx = arange(n, dtype=int)
+    indices = (idx,) * a.ndim
+
+    len_val = n
+
+    if wrap and a.shape[0] > a.shape[1]:
+        len_val = a.shape[0] - int(a.shape[0] / (a.shape[1] + 1))
+
+    if (val.size != len_val and val.ndim > 0) or val.ndim > 1:
+        val = val._wrap(len_val)
+
+    if a.ndim == 2 and wrap and a.shape[0] > a.shape[1]:
+        idx0_tmp = arange(a.shape[1], dtype=int)
+        idx0 = idx0_tmp.copy()
+        while idx0.size < a.shape[0]:
```

Comment:
Should this be `while idx0.size < len_val`?

Reply:
that is right. Thank you